### PR TITLE
Splatfacto in the wild

### PR DESCRIPTION
**Pull request by KevinXu02**
Implimentation of splatfacto in the wild and 2d background model for splatfacto.
https://kevinxu02.github.io/gsw.github.io/

### CODE REVIEW

**Comment by KevinXu02** [comment]
> This PR needs `gsplat==1.0.0` or `gsplat==0.1.12`?

Currently it uses gsplat==0.1.12. I'll update it to gsplat==1.0.0 when I got time.

**Comment by ichsan2895** [comment]
I think it is better If your PR pin specific gsplat and nerfstudio since the latest commit nerfstudio, it already pin `gsplat==1.0.0`.

When it merged to main nerfstudio repo, so it will use `gsplat==1.0.0`.

My idea, pinned your repo with `nerfstudio==1.1.0` and `gsplat==0.1.12`

**Comment by KevinXu02** [comment]
> I think it is better If your PR pin specific gsplat and nerfstudio since the latest commit nerfstudio, it already pin `gsplat==1.0.0`.
> 
> When it merges to main nerfstudio repo, so it will use `gsplat==1.0.0`.
> 
> My idea, pinned your repo with `nerfstudio==1.1.0` and `gsplat==0.1.12`

Thank you for your advice! I guess update this to gsplat==1.0.0 won't be super hard. 1.0 version of gsplat reduced the gram usage a lot thus making this new method can be ran on more devices. And the bg model could also be useful for newer version of nerfstudio.

**Comment by ichsan2895** [comment]
Good luck 👍

**Comment by ichsan2895** [comment]
Good luck 👍

BTW, why the default --steps-per-eval-all-images and --steps-per-eval-image' are set to 1000000` ?

Is there any problem if I lower it?

**Comment by KevinXu02** [comment]
> Good luck 👍
> 
> BTW, why the default `--steps-per-eval-all-images` and `--steps-per-eval-image` are set to 1000000 ?
> 
> Is there any problem if I lower it?

Because the eval for nerf in the wild uses left half eval images for training and another half for eval. I need to sort of hack the dataloader to make it work and it's ugly. It won't be a problem if you lower it, but it will always uses 0th image's appearance embedding for output.

**Comment by KevinXu02** [comment]
> I got error when I exported the splat
> 
> ```
> >>> ns-export gaussian-splat --load-config outputs/unnamed/splatfacto-w/2024-06-10_092617/config.yml --output-dir renders/MY_DATASET_SPLATWILD/
> 
> WARNING: Using slower TCNN CutlassMLP instead of TCNN FullyFusedMLP
> Use layer width of 16, 32, 64, or 128 to use the faster TCNN FullyFusedMLP.
> 
> Loading latest checkpoint from load_dir
> ✅ Done loading checkpoint from outputs/unnamed/splatfacto-w/2024-06-10_092617/nerfstudio_models/step-000064999.ckpt
> Traceback (most recent call last):
>   File "/usr/local/bin/ns-export", line 8, in <module>
>     sys.exit(entrypoint())
>   File "/workspace/NERFSTUDIO_SPLATWILD/nerfstudio/nerfstudio/scripts/exporter.py", line 638, in entrypoint
>     tyro.cli(Commands).main()
>   File "/workspace/NERFSTUDIO_SPLATWILD/nerfstudio/nerfstudio/scripts/exporter.py", line 550, in main
>     assert isinstance(pipeline.model, SplatfactoModel)
> AssertionError
> ```
> 
> additional information
> 
> ```
> RTX A6000 48 GB
> Ubuntu 22.04.3 LTS
> python 3.10.12
> torch 2.0.1+cu118
> https://github.com/KevinXu02/nerfstudio/tree/splatfactow@eb9ee7a0aa7ac31ad768b5a5ac1b7eeae4af10ef
> gsplat==0.1.12
> ```



> I got error when I exported the splat
> 
> ```
> >>> ns-export gaussian-splat --load-config outputs/unnamed/splatfacto-w/2024-06-10_092617/config.yml --output-dir renders/MY_DATASET_SPLATWILD/
> 
> WARNING: Using slower TCNN CutlassMLP instead of TCNN FullyFusedMLP
> Use layer width of 16, 32, 64, or 128 to use the faster TCNN FullyFusedMLP.
> 
> Loading latest checkpoint from load_dir
> ✅ Done loading checkpoint from outputs/unnamed/splatfacto-w/2024-06-10_092617/nerfstudio_models/step-000064999.ckpt
> Traceback (most recent call last):
>   File "/usr/local/bin/ns-export", line 8, in <module>
>     sys.exit(entrypoint())
>   File "/workspace/NERFSTUDIO_SPLATWILD/nerfstudio/nerfstudio/scripts/exporter.py", line 638, in entrypoint
>     tyro.cli(Commands).main()
>   File "/workspace/NERFSTUDIO_SPLATWILD/nerfstudio/nerfstudio/scripts/exporter.py", line 550, in main
>     assert isinstance(pipeline.model, SplatfactoModel)
> AssertionError
> ```
> 
> additional information
> 
> ```
> RTX A6000 48 GB
> Ubuntu 22.04.3 LTS
> python 3.10.12
> torch 2.0.1+cu118
> https://github.com/KevinXu02/nerfstudio/tree/splatfactow@eb9ee7a0aa7ac31ad768b5a5ac1b7eeae4af10ef
> gsplat==0.1.12
> ```



> I got error when I exported the splat
> 
> ```
> >>> ns-export gaussian-splat --load-config outputs/unnamed/splatfacto-w/2024-06-10_092617/config.yml --output-dir renders/MY_DATASET_SPLATWILD/
> 
> WARNING: Using slower TCNN CutlassMLP instead of TCNN FullyFusedMLP
> Use layer width of 16, 32, 64, or 128 to use the faster TCNN FullyFusedMLP.
> 
> Loading latest checkpoint from load_dir
> ✅ Done loading checkpoint from outputs/unnamed/splatfacto-w/2024-06-10_092617/nerfstudio_models/step-000064999.ckpt
> Traceback (most recent call last):
>   File "/usr/local/bin/ns-export", line 8, in <module>
>     sys.exit(entrypoint())
>   File "/workspace/NERFSTUDIO_SPLATWILD/nerfstudio/nerfstudio/scripts/exporter.py", line 638, in entrypoint
>     tyro.cli(Commands).main()
>   File "/workspace/NERFSTUDIO_SPLATWILD/nerfstudio/nerfstudio/scripts/exporter.py", line 550, in main
>     assert isinstance(pipeline.model, SplatfactoModel)
> AssertionError
> ```
> 
> additional information
> 
> ```
> RTX A6000 48 GB
> Ubuntu 22.04.3 LTS
> python 3.10.12
> torch 2.0.1+cu118
> https://github.com/KevinXu02/nerfstudio/tree/splatfactow@eb9ee7a0aa7ac31ad768b5a5ac1b7eeae4af10ef
> gsplat==0.1.12
> ```

I haven't added support for the export yet bc both the gs colors and the background are implicit and need to be specially handled when exporting. But you should be able to view the result in ns-viewer. 

And for splatfacto the gaussians should can be correctly exported but the background should be lost if you enables it. I'm planning to export it as a sky box but the current viewer doesn't support rendering sky boxes.

Thank you for telling me this issue!

**Comment by ichsan2895** [comment]
I think you can create splat for background scene which is shaped like sphere. The color is the interpolation of boundary of current splat. This idea is inspired from Luma AI.

Here is the outside of the sphere
image

Here is the inside of the sphere. I take this scene on (almost) the top of a mountain.
image

**Comment by KevinXu02** [comment]
It supports gsplat 1.0 now and I keep the cull to bigs as the result would looks better.
This also includes a refined background logic.

File: nerfstudio/configs/method_configs.py [hunk 1]
```diff
@@ -637,6 +641,14 @@
                 lr_final=5e-7, max_steps=30000, warmup_steps=1000, lr_pre_warmup=0
             ),
         },
+        "field_background": {
+            "optimizer": AdamOptimizerConfig(lr=4e-3, eps=1e-15),
+            "scheduler": ExponentialDecaySchedulerConfig(lr_final=1e-4, max_steps=30000),
+        },
+        "appearance_embed": {
```
Comment (brentyi) [inline nerfstudio/configs/method_configs.py:648]: Are these being added to normal splatfacto? Any particular reason why?

File: nerfstudio/data/dataparsers/phototourism_dataparser.py [hunk 2]
```diff
@@ -56,6 +58,16 @@ class PhototourismDataParserConfig(DataParserConfig):
     """The method to use to center the poses."""
     auto_scale_poses: bool = True
     """Whether to automatically scale the poses to fit in +/- 1 bounding box."""
+    colmap_path: Path = Path("dense/sparse")
```
Comment (brentyi) [inline nerfstudio/data/dataparsers/phototourism_dataparser.py:61]: maybe note in comment that this is relative to the dataset path

File: nerfstudio/data/dataparsers/phototourism_dataparser.py [hunk 3]
```diff
@@ -176,6 +191,73 @@ def _generate_dataparser_outputs(self, split="train"):
             scene_box=scene_box,
             dataparser_scale=scale_factor,
             dataparser_transform=transform_matrix,
+            metadata=metadata,
         )
 
         return dataparser_outputs
+
+    def _load_3D_points(self, colmap_path: Path, transform_matrix: torch.Tensor, scale_factor: float):
```
Comment (brentyi) [inline nerfstudio/data/dataparsers/phototourism_dataparser.py:199]: are there any differences between `_load_3D_points` here and the version in the colmap dataparser? or is it duplicated?

File: nerfstudio/fields/background_field.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import torch
+from torch import Tensor, nn
+
+from nerfstudio.cameras.rays import RayBundle
+from nerfstudio.field_components import MLP
+from nerfstudio.field_components.encodings import SHEncoding
+from nerfstudio.fields.base_field import Field, get_normalized_directions
+
+
+class BG_Field(Field):
```
Comment (brentyi) [inline nerfstudio/fields/background_field.py:25]: I think the naming convention we use would be `BGField` (eg `SHEncoding`, `RGBRenderer`, etc)

File: nerfstudio/models/splatfacto.py [hunk 5]
```diff
@@ -212,6 +219,7 @@ def populate_modules(self):
         else:
             means = torch.nn.Parameter((torch.rand((self.config.num_random, 3)) - 0.5) * self.config.random_scale)
         self.xys_grad_norm = None
+        self.max_2Dsize = None
```
Comment (brentyi) [inline nerfstudio/models/splatfacto.py:222]: nit: `max_2d_size`?

Comment (brentyi) [reply nerfstudio/models/splatfacto.py:222]: we could also use a more descriptive word here than "size", from my skim I'm not 100% sure what units or what this is

File: nerfstudio/viewer/viewer.py [hunk 6]
```diff
@@ -462,11 +465,16 @@ def init_scene(
...
+            def create_on_click_callback(capture_idx):
+                def on_click_callback(event: viser.SceneNodePointerEvent[viser.CameraFrustumHandle]) -> None:
+                    with event.client.atomic():
+                        event.client.camera.position = event.target.position
+                        event.client.camera.wxyz = event.target.wxyz
+                        self.current_camera_idx = capture_idx
+
+                return on_click_callback
+
+            camera_handle.on_click(create_on_click_callback(idx))
```
Comment (brentyi) [inline nerfstudio/viewer/viewer.py:477]: this also impacts nerfacto, right? could be good in its own PR?

**Review by brentyi — COMMENTED** [review]
I did a pass, awesome work @KevinXu02!! The results look amazing.

I realize it's tough because things depend on each other, but one suggestion is that it would be easier to merge things if we can break this mega-PR into multiple smaller PRs, maybe:
- the viewer appearance embedding changes
- the phototourism dataloader improvements
- 2D background model for splatfacto
- splatfacto-w

In particular I'm hesitant about merging all of splatfacto-w into the core repo for maintenance reasons (as opposed to just putting in place the infra required for splatfacto-w to exist as a plugin/external method[^1]), especially if it's both independent of splatfacto and replicates a lot of logic, since it would increase the burden of a potential splatfacto refactor. But for these decisions I'll defer to splatfacto lead @kerrj :)

[^1]: https://github.com/nerfstudio-project/nerfstudio-method-template

File: nerfstudio/configs/method_configs.py (hunk 1 above)
Comment (KevinXu02) [reply nerfstudio/configs/method_configs.py:648]: Yes. Because enables the bg model in splatfacto would also possibly gives a better outdoor scene.
train
egypt

File: nerfstudio/data/dataparsers/phototourism_dataparser.py (hunk 3 above)
Comment (KevinXu02) [reply nerfstudio/data/dataparsers/phototourism_dataparser.py:199]: Yes this part was copied from the colmap dataparser

File: nerfstudio/viewer/viewer.py (hunk 6 above)
Comment (KevinXu02) [reply nerfstudio/viewer/viewer.py:477]: Basiclly this part will pass the camera idx in camera's metadata when renders a output. I guess it won't be a problem if the model doesn't use the camera metadata?

File: nerfstudio/models/splatfacto.py (hunk 5 above)
Comment (KevinXu02) [reply nerfstudio/models/splatfacto.py:222]: This part belongs to older version of splatfacto that culls screen size too bigs. I add it back in this new one as this makes the results look better.

File: nerfstudio/models/splatfactow.py [hunk 7]
```diff
@@ -0,0 +1,1119 @@
+# ruff: noqa: E741
```
Comment (pwais) [inline nerfstudio/models/splatfactow.py:1]: it would be cool if this project used the nerfstudio-template ( https://github.com/nerfstudio-project/nerfstudio-method-template )

as does e.g. garfield ( https://github.com/chungmin99/garfield ) and the more recent dn_splatter ( https://github.com/maturk/dn-splatter )   

versus copy-pasta the entire splatfacto impl and hack stuff up just for one paper (as is done in this file).  it appears there are indeed a few parts of nerfstudio that need changing tho. highly recommend start as a template project and then maybe later port into main nerfstudio as newest splatfacto.  

while i do think down the line any gsplat user probably wants a good model that has a background model, should users really get pre-print material? or refereed, maybe even journal-quality, i.e. "de-facto".

Comment (KevinXu02) [reply nerfstudio/models/splatfactow.py:1]: Yes I’m working on making this a template to make everything clearer. And probably need more tests on the viser side to make sure it won’t hurt other models.
Actually the first version of this works as a plugin of splatfacto but its implicit nature looks conflicted with the original gs. This makes everything messy.
As for the material wise for background model, I’m not very sure about if I should release the paper along with the code because it’s still unfinished and this is actually the first time I’ve tried to do a research project.
Thank you so much for your suggestion!

**Comment by kerrj** [comment]
Closing this since @KevinXu02  is planning on making a template repo!
